### ISSUE TITLE
Support for $ref tags/JSON schema in extension's package.json

### ISSUE THREAD

**Issue by bryceitoc9** [microsoft/vscode#95270]
I work on a [VS Code Extension](https://github.com/aws/aws-toolkit-vscode), and I'm following up on https://github.com/microsoft/vscode/issues/95033 . It looks like this issue was caused by the fact that we were using a [`$ref` field](https://github.com/aws/aws-toolkit-vscode/blob/v1.8.0/package.json#L247) within the package.json manifest. This appears to be an invalid use of package.json, given the unusual behavior shown in #95033. Could you please explain whether or not this is valid in package.json?

To offer some insight, we have embedded a JSON Schema into the `debugger` contribution point. The motivation was to use a [code generator](https://github.com/aws/aws-toolkit-vscode/blob/v1.8.0/build-scripts/generateConfigurationAttributes.ts) against the JSON schema to ensure the extension code remains in sync with the structure defined in the manifest. If this scenario is not valid, we'd be interested in hearing a recommended way to keep the manifest and code structures in sync.

Thanks in advance!

**Comment by weinand** [comment]
@isidorn @aeschli we are combining the json schemas from all debuggers into a single schema and then use this for the launch.json editors. Are there any limitations what this schema can use? Or are we merging the fragments in a problematic way?

**Comment by isidorn** [comment]
Well the issue is probably due to the `$ref` field as @bryceitoc9 suggests.
@aeschli do you treat this $ref special in the json schema land? Was it always like this?

**Comment by aeschli** [comment]
$ref can not be used in such schema fragment as the path is resolved against the composed schema that users don't know how it looks like and should not know.
Maybe in the future when we support using $id in $ref (https://github.com/microsoft/vscode/issues/92042) this can be allowed.

**Comment by aeschli** [comment]
Not a bug but a feature request.

**Comment by awschristou** [comment]
It sounds like there is a restriction to not have `$ref` fields (and maybe embedded json schemas?) within the package.json file. Is this documented somewhere?

**Comment by aeschli** [comment]
@weinand This needs to be documented with the `debuggers` contributions point.
We already do the same for the `configuration` contributions point: https://code.visualstudio.com/api/references/contribution-points#Configuration-property-schema
(implicitly, be describing what is supported)

**Comment by matepek** [comment]
> (implicitly, by describing what is supported)

From the site: https://code.visualstudio.com/api/references/contribution-points#Configuration-property-schema

>> Configuration keys are defined using a superset of JSON Schema.
>
>> You can use any the properties defined by JSON Schema to describe other constraints on configuration values.

Without ref I cannot define a recursive schema. 

Do I miss something?😏

**Comment by aeschli** [comment]
You can only use other properties that define a constraint.

I made this more explicit: https://github.com/microsoft/vscode-docs/commit/7108f817586e28808f969c4fb517d2b9b51ee001

**Comment by bwateratmsft** [comment]
The Docker extension has a substantial amount of repetition in its package.json; this would be a very nice feature to have.

**Comment by bwateratmsft** [comment]
Update since some folks might be interested. I found a rather hacky way to accomplish this--I tried it for settings but it may also work for tasks/launch configs as well. Please note, it only helps with the settings JSON editor, not in the settings UI (which will simply send you to the JSON).

You can use `"$ref"` if you feed it an absolute URI to a schemas file, but the scheme for that URI can be anything. It seems internally VSCode uses `vscode://`, for example. By registering a [FileSystemProvider](https://code.visualstudio.com/api/references/vscode-api#FileSystemProvider) with a custom scheme (let's say, `myschemas://`), I was able to reference schema files from package.json and get working validation / autocomplete / etc.

Basically what I did is:
1. Add an activation event in package.json, `"onFileSystem:myschemas"`
1. Register said filesystem provider: `vscode.workspace.registerFileSystemProvider("myschemas", new MySchemasFilesystemProvider())`
1. Implement just `stat` and `readFile` on `MySchemasFilesystemProvider`, using the extension install directory as the root (so your schemas.json file can be adjacent to your package.json)
1. Have settings that use `"$ref": "myschemas:///schemas.json#/definitions/mycomplexobjectdefn"`

If there is already some built-in `FileSystemProvider` for getting at extension files, steps 1-3 could be skipped, but I'm not aware of one.
